js/ui/windowManager.js

```diff
@@ -2166,3 +1888,368 @@ const WindowManager = new Lang.Class({
+            });
+            session.windowBuilder.disconnect(session.windowBuilder._firstFrameConnection);
+            session.windowBuilder._firstFrameConnection = null;
+            this._addButton(session);
+            session.buttonBuilder.show();
+            session.buttonApp.hide();
+            return false;
+        }));
+        /* Save the connection's id on the destination window and in a list too so we can
+         * get rid of it on kill-window-effects later */
+        session.windowBuilder._firstFrameConnection = firstFrameConnection;
+        this._firstFrameConnections.push(firstFrameConnection);
+    },
+
+    _windowAppPositionChanged: function(window, event, session) {
+        let rect = session.windowApp.get_meta_window().get_frame_rect();
+        session.buttonApp.set_position(rect.x + rect.width - BUTTON_OFFSET_X, rect.y + rect.height - BUTTON_OFFSET_Y);
+        if (!session.windowBuilder)
+                return;
+        session.windowBuilder.get_meta_window().move_resize_frame(false,
+                                                                  rect.x,
+                                                                  rect.y,
+                                                                  rect.width,
+                                                                  rect.height);
+    },
+
+    _windowBuilderPositionChanged: function(window, event, session) {
+        let rect = session.windowBuilder.get_meta_window().get_frame_rect();
+        session.buttonBuilder.set_position(rect.x + rect.width - BUTTON_OFFSET_X, rect.y + rect.height - BUTTON_OFFSET_Y);
+        if (!session.windowApp)
+            return;
+        session.windowApp.get_meta_window().move_resize_frame(false,
+                                                              rect.x,
+                                                              rect.y,
+                                                              rect.width,
+                                                              rect.height);
+    },
+
+    _windowAppRestacked : function(overview, stackIndices, session) {
+        let focusedWindow = global.display.get_focus_window();
+        if (!focusedWindow)
+            return;
+        if (!session.windowApp.get_meta_window())
+            return;
+
+        if (session.buttonApp)
+            session.buttonApp.hide();
+        if (session.buttonBuilder)
+            session.buttonBuilder.hide();
+
+        if (focusedWindow.get_stable_sequence() === session.windowApp.get_meta_window().get_stable_sequence()){
+            if (!session.state) {
+                session.buttonApp.show();
+                return;
+            }
+            if (!session.windowBuilder)
+                return;
+
+            session.windowBuilder.show();
+            session.buttonBuilder.show();
+            session.windowBuilder.get_meta_window().activate(global.get_current_time());
```

Comment:
This does the handling of the case where we switch to the session from another app using alt-tab for example. If we have the Builder window as current "view" before we will be switching back to it.